trace-recorder/src/main/org/jetbrains/lincheck/trace/recorder/TraceCollectingEventTracker.kt

```diff
@@ -616,13 +681,48 @@ class TraceCollectingEventTracker(
         startTime = System.currentTimeMillis()
     }
 
+    fun finishRunningThread(thread: Thread) {
+        val threadDescriptor = ThreadDescriptor.getThreadDescriptor(thread) ?: return
+        val threadData = threadDescriptor.eventTrackerData as? ThreadData? ?: return
+        if (!threadDescriptor.isAnalysisEnabled) return // we skip already finished threads
+
+        // tell that thread not to track any tracepoints anymore
+        threadDescriptor.disableAnalysis()
+        // wait until that thread finishes whatever injected code it is executing right now
+        spinner.spinWaitUntil { !threadDescriptor.isInsideInjectedCode }
+        // now, we are sure that another thread has finished its injected code
+        // and will not attempt to execute anything else, because we disabled analysis in it
+        val stackFrames = threadData.getStack().asReversed()
+        stackFrames.forEach { frame ->
+            val tracePoint = frame.call
+            tracePoint.result = TR_OBJECT_UNFINISHED_METHOD_RESULT
+            strategy.callEnded(thread, tracePoint)
+        }
+        strategy.completeThread(thread)
+    }
+
     fun finishAndDumpTrace() {
+        // TODO:
+        //  1. disable analysis in all still-running threads
+        //     a. finish threads via CAS, so `afterThreadFinish` and `finishRunningThread` are not called simultaneously
+        //  2. we should write the ends of footers of each still-open MethodCallTracePoint's
+        //  3. - mark them some-how that their body wasn't finished when main finished
+        //  4. collect the prefix of the existing trace and create trace points for that
+        //  5. - mark existing threads that their prefixes are missing
+
+        // Finish existing threads, except for Main
+        val mainThread = Thread.currentThread()
+        threads.forEach { (thread, _) ->
+            if (thread != mainThread) {
+                finishRunningThread(thread)
```

Comment:
Probably better to first call `disableAnalysis()` for all threads, and then wait for their completion.